### PR TITLE
Exploring GraphiQL: The In-Browser IDE for GraphQL

### LINKED ISSUE #375: Write a Post Outline for Exploring GraphiQL: The In-Browser IDE for GraphQL

### PR DESCRIPTION
**Pull request by shegz101**
This guide gives readers an overview of GraphiQL, an in-browser IDE for GraphQL that allows users to run queries and see the responses.

fixes #375 
/claim #375

### CODE REVIEW

File: outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
+# Exploring GraphiQL: The In-Browser IDE for GraphQL
+
+**Target Audience:** This article is for all developers and testers who integrate or test GraphQL APIs, whether front-end developers, back-end developers, or QA testers. It's also for anyone else interested in the world of GraphQL APIs.
+
+**Aim:** This article aims to provide developers with an in-depth understanding of GraphiQL, its features, and how to effectively use it to interact with GraphQL APIs. It will cover the fundamentals, and practical applications for utilizing GraphiQL as an in-browser IDE for GraphQL.
+
+## Introduction
+Why waste time and effort troubleshooting your GraphQL API directly in your application when there's a more efficient way? Imagine testing and perfecting your queries outside your app, ensuring they work flawlessly before implementation. Sounds ideal, right? With [GraphiQL](https://az.dev/swapi-graphql), you're saving time and gaining control over your development process, empowering you to work more confidently and effectively.
```
Comment (amitksingh1490) [inline outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md]:
```suggestion
Why waste time and effort troubleshooting your GraphQL API directly in your application when there's a more efficient way? Imagine testing and perfecting your queries outside your app, ensuring they work flawlessly before implementation. Sounds ideal, right? With [GraphiQL](https://tailcall.run/playground/), you're saving time and gaining control over your development process, empowering you to work more confidently and effectively.
```

File: outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+## Why use GraphiQL?
+* Advantages/benefits of using GraphiQL (things like Improving your request, validating your improvements, and easily debugging requests running into problems)
+
+## Methods for accessing GraphiQL
+To access GraphiQL, you are not limited to the default in-browser method. 
+* Head to [https://graphql.org/swapi-graphql](https://graphql.org/swapi-graphql) by querying your search engine (browser) to access the GraphiQL online editor, which works with the Star Wars data, and it is publicly available for you to use
+* If you are using Gatsby, you can start the Gatsby development server by pasting this: `http://localhost:8000/___graphql` into your browser
+* Using Shopify via the [Shopify GraphiQl app](https://shopify-graphiql-app.shopifycloud.com/login)
+* Using your own API endpoint: just type your GraphQL server endpoint URL into your browser
```
Comment (amitksingh1490) [inline outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md]: use https://tailcall.run/playground/

File: outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+## Getting Started with the GraphiQL
+* Configuring GraphiQL
+* Connecting GraphiQL to your GraphQL server
```
Comment (amitksingh1490) [inline outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md]:
```suggestion
* Connecting GraphiQL to your Tailcall GraphQL server
```

File: outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+* Exploring the GraphiQL interface and understanding the functions of the various panes. The application is two-pane by default, but it becomes three when we access the Documentation Explorer.
+* Explain each of the main parts in the GraphiQL interface: Query Editor, Variables Editor, Response pane, and Documentation Explorer.
+
+## Key Features and Capabilities of GraphiQl
+Explain the various core features that GraphiQL provides.
+* Type-ahead and Auto Completion feature that is aware of the GraphQL type schema you are currently exploring. In addition, explain Context-awareness.
+* Live syntax and validation error highlighting.
+* Query history
+* Fragments: This feature helps us maintain the Don’t Repeat Yourself (DRY) pattern when dealing with repetitive code in our query. Instead, you can easily merge fragments into a query.
+* Documentation: GraphiQL generates documentation for your schema, making it easy to understand and explore.
```
Comment (amitksingh1490) [inline outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md]: screenshot for documentation How it looks

File: outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+## Testing out the Interface
+Writing and executing Queries, mutations and Subscriptions
+* Writing the query
+* Integrating Variables with Your Query
+* Utilising fragments to aid query construction
+* Executing/Testing the query: Sending requests and handling responses
```
Comment (amitksingh1490) [inline outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md]: screen shots

File: outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+## What is GraphiQL?
+* Definition and overview of GraphiQL
+
+## Why use GraphiQL?
+* Advantages/benefits of using GraphiQL (things like Improving your request, validating your improvements, and easily debugging requests running into problems)
+
+## Methods for accessing GraphiQL
+To access GraphiQL, you are not limited to the default in-browser method. 
+* Head to [https://graphql.org/swapi-graphql](https://graphql.org/swapi-graphql) by querying your search engine (browser) to access the GraphiQL online editor, which works with the Star Wars data, and it is publicly available for you to use
+* If you are using Gatsby, you can start the Gatsby development server by pasting this: `http://localhost:8000/___graphql` into your browser
```
Comment (amitksingh1490) [inline outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md]:
```suggestion
* If you are using Gatsby, you can start the Gatsby development server by pasting this: `http://localhost:8000/___graphql` into your browser
```
This should reference tailcall

Comment (shegz101) [reply outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md]: @amitksingh1490 How do I do this? Is it the Tailcall website I am referencing?

Comment (amitksingh1490) [reply outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md]: check this out https://tailcall.run/docs/#starting-the-graphql-server

File: outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+## What is GraphiQL?
+* Definition and overview of GraphiQL
+
+## Why use GraphiQL?
+* Advantages/benefits of using GraphiQL (things like Improving your request, validating your improvements, and easily debugging requests running into problems)
+
+## Methods for accessing GraphiQL
+To access GraphiQL, you are not limited to the default in-browser method. 
+* Head to [https://tailcall.run/playground/?u=http://127.0.0.1:8000/graphql](https://tailcall.run/playground/?u=http://127.0.0.1:8000/graphql) by querying your search engine (browser) to access the GraphiQL online editor, which works with the Star Wars data, and it is publicly available for you to use
+* If you are using Gatsby, you can start the Gatsby development server by referencing this: (https://tailcall.run/docs/#starting-the-graphql-server)[https://tailcall.run/docs/#starting-the-graphql-server]
```
Comment (amitksingh1490) [inline outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md]: > start the Gatsby development server

Doesn't make sense

Comment (shegz101) [reply outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md]: Alright. But is the first point okay?

Comment (shegz101) [reply outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md]: > > start the Gatsby development server
> 
> Doesn't make sense

Alright. I'd just remove that and add the part of starting a server from the link you shared earlier.

Comment (amitksingh1490) [reply outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md]: >which works with the Star Wars data, and it is publicly available for you to use 

is incorrect

Comment (shegz101) [reply outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md]: I'd just remove that part. The initial link supports the Star Wars data.

Comment (shegz101) [reply outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md]: > > which works with the Star Wars data, and it is publicly available for you to use
> 
> is incorrect

To clarify is there more than one method to access GraphiQL when using Tailcall?

Comment (amitksingh1490) [reply outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md]: when you start the Tailcall server you get a link to playground.
or you can directly go to the https://tailcall.run/playground/ and give your graphql server url in input
Only thing you need to make sure your graphql server should allow cors for Tailcall domain

Comment (shegz101) [reply outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md]: > when you start the Tailcall server you get a link to playground. or you can directly go to the https://tailcall.run/playground/ and give your graphql server url in input Only thing you need to make sure your graphql server should allow cors for Tailcall domain

Alright, now I get it. Thank you for clarifying. I have committed the changes.

File: outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+
+B. Add a table to compare GraphiQl to Altair Client and Insomnia
+
+## Conclusion
+
+A. Summary of what we discussed
+
+B. Final thoughts on the subject
+
+## Resources
```
Comment (amitksingh1490) [inline outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md]: Drop this section

File: outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+
+B. Using modular queries, i.e. breaking down complex queries into smaller and manageable chunks.
+
+C. Implement thorough/consistent query testing strategies to detect errors early.
+
+D. Adding in-line comments in your Query
+
+## Alternatives to GraphiQL
+
+A. Explain that there are also alternatives to GraphiQL. There are other GraphQL playgrounds like Insomnia, Altair client and so on
```
Comment (amitksingh1490) [inline outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md:70]: add screenshots for each alternative

File: outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+
+B. Error Messages
+
+- Understanding and interpreting GraphQL error messages
+
+C. Network Tab in Developer Tools
+
+- Analyzing network requests and responses for additional insights.
+
+### Best Practices
```
Comment (amitksingh1490) [inline outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md]: Not relavant

File: outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+
+A. Exploring the GraphiQL interface and understanding the functions of the various panes.
+
+B. Explain each of the main parts in the GraphiQL interface
+
+- Query Editor
+- Variables Editor
+- Response pane
+- Documentation Explorer.
+
```
Comment (amitksingh1490) [inline outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md]: How to add headers?

**Comment by amitksingh1490** [comment]
Fixes #375

File: outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md (hunk 8 above)
Comment (shegz101) [reply outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md]: Alright, I will.

Comment (shegz101) [reply outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md]: > Drop this section

Including the conclusion?

File: outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md (hunk 10 above)
Comment (shegz101) [reply outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md]: Alright, I'd remove the whole thing.

File: outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md (hunk 8 above)
Comment (amitksingh1490) [reply outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md]: ## Resources section

File: outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md (hunk 11 above)
Comment (shegz101) [reply outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md]: Do you mean the parts I added "head to"?

File: outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md (hunk 8 above)
Comment (shegz101) [reply outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md]: > ## Resources section

Done

File: outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md (hunk 10 above)
Comment (shegz101) [reply outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md]: > Not relavant

Here, I removed all the sections from "Key Features and Capabilities of GraphiQl" to "Debugging Strategies"

File: outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md (hunk 11 above)
Comment (amitksingh1490) [reply outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md]: How to add headers? in GraphiQL

File: outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md (hunk 8 above)
Comment (amitksingh1490) [reply outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md]: its still there

**Review by amitksingh1490 — CHANGES_REQUESTED** [review]

File: outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md (hunk 11 above)
Comment (shegz101) [reply outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md]: Alright, I'd add that.

File: outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md (hunk 8 above)
Comment (shegz101) [reply outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md]: > its still there

I have resolved it in this new commit. Thank you!

**Review by amitksingh1490 — APPROVED** [review]

**Pull request merged**